### PR TITLE
feat: Upload C2 supporting documents (FPLA-2131)

### PR DESCRIPTION
**Pull request by KiranVarma02**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-2131

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: e2e/pages/events/uploadC2DocumentsEvent.page.js
Comment (Jamesnelson117) [inline e2e/pages/events/uploadC2DocumentsEvent.page.js:40]: Should be part of the test rather than the page in my opinion.

File: ccd-definition/ComplexTypes/CareSupervision/1_ManageDocuments/ManageDocumentBundle.json
Comment (Jamesnelson117) [inline ccd-definition/ComplexTypes/CareSupervision/1_ManageDocuments/ManageDocumentBundle.json:23]: Do we need any validation in the UI?

File: ccd-definition/ComplexTypes/CareSupervision/1_ManageDocuments/ManageDocumentBundle.json
Comment (KiranVarma02) [reply ccd-definition/ComplexTypes/CareSupervision/1_ManageDocuments/ManageDocumentBundle.json:4]: this ticket context is supportingDocuments but something generic is fine i guess!

File: e2e/pages/events/uploadC2DocumentsEvent.page.js
Comment (KiranVarma02) [reply e2e/pages/events/uploadC2DocumentsEvent.page.js:40]: i think i agree with you. I just followed same approach as other tests but happy to move all... what do you think?

File: ccd-definition/ComplexTypes/CareSupervision/1_ManageDocuments/ManageDocumentBundle.json
Comment (KiranVarma02) [reply ccd-definition/ComplexTypes/CareSupervision/1_ManageDocuments/ManageDocumentBundle.json:23]: added it

**Comment by Jamesnelson117** [comment]
@KiranVarma02 are you sure this is ready? What about 'date and time uploaded' (see 2131 designs screen on Jira) and toggle?

File: ccd-definition/ComplexTypes/CareSupervision/1_ManageDocuments/ManageDocumentBundle.json
Comment (Jamesnelson117) [reply ccd-definition/ComplexTypes/CareSupervision/1_ManageDocuments/ManageDocumentBundle.json:23]: Where?

File: e2e/pages/events/uploadC2DocumentsEvent.page.js
Comment (Jamesnelson117) [inline e2e/pages/events/uploadC2DocumentsEvent.page.js:25]: This might be nicer having 'supportingDocument' or similar as a parent key. Then we could have

'SupportingDocument': {
  'name': '',
  'notes': ''
  etc etc
}

File: ccd-definition/ComplexTypes/CareSupervision/1_ManageDocuments/ManageDocumentBundle.json
Comment (KiranVarma02) [reply ccd-definition/ComplexTypes/CareSupervision/1_ManageDocuments/ManageDocumentBundle.json:23]: ahh you mean validation user while entering !! i was missing date assertion before and i thought you were talking about that and added it. 

No validation mentioned in the ticket but i will check with Gail.

File: e2e/pages/events/uploadC2DocumentsEvent.page.js
Comment (KiranVarma02) [reply e2e/pages/events/uploadC2DocumentsEvent.page.js:25]: done

File: e2e/pages/events/uploadC2DocumentsEvent.page.js
Comment (KiranVarma02) [reply e2e/pages/events/uploadC2DocumentsEvent.page.js:38]: done

File: e2e/pages/events/uploadC2DocumentsEvent.page.js
Comment (KiranVarma02) [reply e2e/pages/events/uploadC2DocumentsEvent.page.js:25]: done

File: ccd-definition/CaseEventToFields/CareSupervision/uploadC2.json
Comment (KiranVarma02) [inline ccd-definition/CaseEventToFields/CareSupervision/uploadC2.json:27]: I have to remove this file, isn't it?

File: ccd-definition/ComplexTypes/CareSupervision/1_ManageDocuments/ManageDocumentBundle.json
Comment (KiranVarma02) [reply ccd-definition/ComplexTypes/CareSupervision/1_ManageDocuments/ManageDocumentBundle.json]: validation added

File: ccd-definition/CaseEventToFields/CareSupervision/uploadC2.json
Comment (akomisarek) [reply ccd-definition/CaseEventToFields/CareSupervision/uploadC2.json:27]: Probably yes, or rename it to -prod? Depending how you want to tackle it. Just reviewing!

File: ccd-definition/CaseEventToFields/CareSupervision/uploadC2-prod.json
Comment (akomisarek) [inline ccd-definition/CaseEventToFields/CareSupervision/uploadC2-prod.json:12]: So by looking at the controller it seems you renamed that URL, this means the code you are pushing for prod is going to fail as there is no longer this endpoint. You either need to have two endpoints, or implement feature toggle to switch the code, or implement the code which would work for two flows. 

Or am I missing something? But I don't see this endpoint anymore... Remember merging the code pushes that code to prod.

File: ccd-definition/CaseEventToFields/CareSupervision/uploadC2.json
Comment (akomisarek) [reply ccd-definition/CaseEventToFields/CareSupervision/uploadC2.json:27]: Seems like you need to remove yes, but remember backend needs to support two ways 🤔

File: ccd-definition/ComplexTypes/CareSupervision/1_SupportingDocuments/SupportingEvidenceBundle-nonprod.json
Comment (akomisarek) [inline ccd-definition/ComplexTypes/CareSupervision/1_SupportingDocuments/SupportingEvidenceBundle-nonprod.json:1]: I think this file doesn't need to be in "nonprod" section. Even if we configure it in prod nothing wrong is going to happen.

File: e2e/pages/events/uploadC2DocumentsEvent.page.js
Comment (akomisarek) [reply e2e/pages/events/uploadC2DocumentsEvent.page.js:43]: Don't have opinion here - whatever works better for you.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsAboutToSubmitControllerTest.java
Comment (akomisarek) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsAboutToSubmitControllerTest.java:98]: What if the returned data is empty? 
What if I break the code and don't return the bundle, whatever test is using that, it's still oing to pass.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java:93]: As mentioned in the defs file, how this going to work on prod? Is the code below going to run on the old data?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java:165]: How this is tested? That appropriate date is added to the document? 

I'ts not much logic above, but this probably should live in the service.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleServiceTest.java
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleServiceTest.java:37]: I would add another edge case - Today ;)

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/UploadC2DocumentsService.java
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/UploadC2DocumentsService.java:16]: I am not biggest fan of this class and below, I think they are tightly coupled and there is not massive benefit of having them separate and wiring through the dependency injection (unless this one has more functionality as stated in the comment above in the controller. 

It lacks unit test as well.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/UploadC2DocumentsService.java
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/UploadC2DocumentsService.java:23]: Is this filtering required? Empty list should pass the validation.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java:18]: I am not sure if reusing some group is good thing here, are we sure it's the same meaning as that group should denote? Alternatively I am not sure if you need even group at all - you are just validating single class, do you need to pass any group? 

Seems a bit confusing to me!

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java:16]: Does it need to be more complex by introducing the generics? 

The code seems to work with List<SupportingEvidenceBundle>

**Review by akomisarek — CHANGES_REQUESTED** [review]

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java
Comment (KiranVarma02) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java:16]: @Jamesnelson117 is going to use this for all other supporting documents.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java
Comment (KiranVarma02) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java:18]: not needed, you are right. Removed.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleServiceTest.java
Comment (KiranVarma02) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleServiceTest.java:37]: done

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/UploadC2DocumentsService.java
Comment (KiranVarma02) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/UploadC2DocumentsService.java:23]: not needed. Removed.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/UploadC2DocumentsService.java
Comment (KiranVarma02) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/UploadC2DocumentsService.java:16]: added unit test. Idea is for @Jamesnelson117 to use `ValidateSupportingEvidenceBundleService` for other supporting evidences he is working on ...

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java
Comment (KiranVarma02) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java:165]: I hope you know how much i wanted to move this to service 😄 , there wasn't a service at all when i was working on this... now i added it to do validation. I don't want to put just my code in this PR to service which is bad(i think), so i have to move to all private method which would mean adding tests to things not related to story. If i get time, I would love to do it. Don't want to hold up @Jamesnelson117 . I promise, If i don't do this before this PR is merged, I will do follow up PR 😄 .

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java
Comment (KiranVarma02) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java:93]: I am not 100% sure here, please help me to understand this better. Users can't edit UploadedC2 at the moment, so there is no scenario for validating old data, is there?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsAboutToSubmitControllerTest.java
Comment (KiranVarma02) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsAboutToSubmitControllerTest.java:98]: First test tests supportingEvidenceBundle along with other c2 related data. 

Second test is using different test data to test additional C2 documents and supportingBundle can be empty. 

Seems okay to me.. Let me know what you think.

File: e2e/pages/events/uploadC2DocumentsEvent.page.js
Comment (KiranVarma02) [reply e2e/pages/events/uploadC2DocumentsEvent.page.js:43]: I am going to leave this for now, i think!

File: ccd-definition/ComplexTypes/CareSupervision/1_SupportingDocuments/SupportingEvidenceBundle-nonprod.json
Comment (KiranVarma02) [reply ccd-definition/ComplexTypes/CareSupervision/1_SupportingDocuments/SupportingEvidenceBundle-nonprod.json:1]: agree, done.

File: ccd-definition/CaseEventToFields/CareSupervision/uploadC2.json
Comment (KiranVarma02) [reply ccd-definition/CaseEventToFields/CareSupervision/uploadC2.json:27]: So, l reintroduced `validate-pba-number/mid-event` backend code. I will test everything again but should work now.

File: ccd-definition/CaseEventToFields/CareSupervision/uploadC2-prod.json
Comment (KiranVarma02) [reply ccd-definition/CaseEventToFields/CareSupervision/uploadC2-prod.json:12]: you are right, I should not delete this backend code. I have put that back.

File: e2e/pages/events/uploadC2DocumentsEvent.page.js
Comment (Jamesnelson117) [inline e2e/pages/events/uploadC2DocumentsEvent.page.js:28]: I think since this is a CCD collection, your e2e page object should support the ability of adding multiple. Even if you don't make use of it yourself it can potentially help others down the line.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsAboutToSubmitControllerTest.java
Comment (akomisarek) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsAboutToSubmitControllerTest.java:98]: The problem is you can break the code and this code is still passing. I commented out:

Line 171 in the UploadC2DocumentsController:
//            .supportingEvidenceBundle(wrapElements(updatedSupportingEvidenceBundle));

This test is not enforcing the value to be there, it passes even if nothing is returned. That's for sure wrong.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java:177]: No that's fine, we can do it later! Feel free to create the ticket please so we don't forget :).

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/UploadC2DocumentsService.java
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/UploadC2DocumentsService.java:16]: Ah ok, cool that makes sense then.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java:15]: Ok, probably he should refactor then as we don't know the final design, but it's fine. Happy to leave it here!

Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java:15]: @akomisarek @KiranVarma02 no need for generics here. List is always going to be `List<Element<SupportingEvidenceDocument>>`

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java:94]: I think in this instance it would be beneficial to leave TODO / FIXME / comment stating why we need to have two similar methods. It's confusing when looking at it.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java:15]: Good :)

**Review by akomisarek — CHANGES_REQUESTED** [review]

File: e2e/pages/events/uploadC2DocumentsEvent.page.js
Comment (lewisbirks) [inline e2e/pages/events/uploadC2DocumentsEvent.page.js:46]: Note that there is a helper method that offers part of this functionality, `I.fillDate(...)`. Could probably be extended to fill in the time as well.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java:94]: Hmm, what do you mean sorry @akomisarek? Could we not just have

```
@Postmapping("validate-supporting-evidence-document)"
public AboutToStartOrSubmitCallbackResponse handleValidateSupportingEvidenceDocument
```

File: ccd-definition/ComplexTypes/CareSupervision/1_SupportingDocuments/SupportingEvidenceBundle.json
Comment (lewisbirks) [inline ccd-definition/ComplexTypes/CareSupervision/1_SupportingDocuments/SupportingEvidenceBundle.json:8]: Do we still need this TODO?

File: e2e/pages/events/uploadC2DocumentsEvent.page.js
Comment (lewisbirks) [reply e2e/pages/events/uploadC2DocumentsEvent.page.js:28]: `supportingDocuments` can be turned into a function which takes an index, this index can be then used in the place of `0`, e.g.
```js
    supportingDocuments: (index) => ({
      name: `#temporaryC2Document_supportingEvidenceBundle_${index}_name`,
    }),

```

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsAboutToSubmitControllerTest.java
Comment (KiranVarma02) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsAboutToSubmitControllerTest.java:98]: Sorry, I am not sure if i understand why test should enforce value to be there. SupportingEvidence can be empty. 

I have updated test now and added supportingEvidence to second test too and removed if condition in the test.. not sure if this is any better!

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java
Comment (KiranVarma02) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsController.java:94]: @akomisarek meant to say, now in this PR, PBA validation is done twice once old way and once in generic endpoint. So, I have put TODO to remove old PBA validation.

File: ccd-definition/ComplexTypes/CareSupervision/1_SupportingDocuments/SupportingEvidenceBundle.json
Comment (KiranVarma02) [reply ccd-definition/ComplexTypes/CareSupervision/1_SupportingDocuments/SupportingEvidenceBundle.json:8]: @Jamesnelson117 - You were looking at this - What do we do? Shall i leave it for now and you tackle in your PR?

File: e2e/pages/events/uploadC2DocumentsEvent.page.js
Comment (KiranVarma02) [reply e2e/pages/events/uploadC2DocumentsEvent.page.js:28]: I tried exactly that before you added your comment but it didn't work and reverted back to look at later. Let me try again!

Comment (lewisbirks) [reply e2e/pages/events/uploadC2DocumentsEvent.page.js:28]: it was just a guess, but I know that we have done it else where

Comment (lewisbirks) [reply e2e/pages/events/uploadC2DocumentsEvent.page.js:28]: See [sendCaseToGatekeeperEvent.page.js](https://github.com/hmcts/fpl-ccd-configuration/tree/master/e2e/pages/events/sendCaseToGatekeeperEvent.page.js)

Comment (KiranVarma02) [reply e2e/pages/events/uploadC2DocumentsEvent.page.js]: done

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java
Comment (KiranVarma02) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/ValidateSupportingEvidenceBundleService.java:15]: ok, removing Generic.

File: ccd-definition/ComplexTypes/CareSupervision/1_SupportingDocuments/SupportingEvidenceBundle.json
Comment (Jamesnelson117) [reply ccd-definition/ComplexTypes/CareSupervision/1_SupportingDocuments/SupportingEvidenceBundle.json:8]: TODO was just a placeholder in my FPLA-2001. Remove.

File: e2e/pages/events/uploadC2DocumentsEvent.page.js
Comment (KiranVarma02) [reply e2e/pages/events/uploadC2DocumentsEvent.page.js]: done

File: e2e/steps_file.js
Comment (akomisarek) [inline e2e/steps_file.js:174]: Any reason why + 1 is done here? I guess it's because we use 0-based month in the setup file? couldn't we use just normal 1:1 month? seems a bit weird!

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsAboutToSubmitControllerTest.java
Comment (akomisarek) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsAboutToSubmitControllerTest.java:106]: This is not ideal, as in case of the system not properly replying it will file with indexoutofbounds exception.

I would expect explicit assertion ensuring there is one element before reaching for it. But it's good enough :)

**Review by akomisarek — APPROVED** [review]

File: e2e/steps_file.js
Comment (lewisbirks) [reply e2e/steps_file.js:174]: Date data type uses 0 based month indexes, when we pass our custom date object then it is, or at least should be, 1 based index

Comment (KiranVarma02) [reply e2e/steps_file.js:174]: yes, its 0-based month - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getMonth

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsAboutToSubmitControllerTest.java
Comment (KiranVarma02) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/UploadC2DocumentsAboutToSubmitControllerTest.java]: added check for not empty

**Review by lewisbirks — APPROVED** [review]

File: e2e/tests/hmctsAdministersCaseAfterSubmission_test.js
Comment (Jamesnelson117) [inline e2e/tests/hmctsAdministersCaseAfterSubmission_test.js:118]: Hmm what about document date and time uploaded?

**Review by lewisbirks — DISMISSED** [review]

**Review by lewisbirks — DISMISSED** [review]

**Review by lewisbirks — APPROVED** [review]

**Pull request merged**
